server/src/main/java/org/elasticsearch/search/DefaultSearchContext.java

```diff
@@ -302,6 +302,11 @@ public Query buildFilteredQuery(Query query) {
 
     private Query createTypeFilter(String[] types) {
         if (types != null && types.length >= 1) {
+            if (Arrays.asList(types).contains(MapperService.SINGLE_MAPPING_NAME)) {
```

Comment:
I think this was added to deal with 'explain' requests, but for me this makes the overall search behavior a bit unintuitive. If a user explicitly adds a (deprecated) type filter on `_doc` in a search, I think they would expect the filter to be applied instead of being ignored. From a user's perspective the search APIs differ from the document CRUD APIs, where the `_doc` is seen as part of the endpoint and not an explicit filter.